drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,632 @@
+
+	/* if top is not 2^n-1 */
+	return (val >= old) ? (val - old) : val + top + 1 - old;
+}
+
+static void set_cc_int_pending(const struct device *dev, uint8_t chan)
+{
+	atomic_or(&get_dev_data(dev)->cc_int_pending, BIT(chan));
+	NVIC_SetPendingIRQ(get_stm32_config(dev)->irqn);
+}
+
+static int set_cc(const struct device *dev, uint8_t id, uint32_t val,
+		  uint32_t flags)
+{
+	__ASSERT_NO_MSG(get_dev_data(dev)->guard_period < get_top_value(dev));
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool irq_on_late;
+	TIM_TypeDef *timer = get_stm32_config(dev)->timer;
+	uint32_t top = get_top_value(dev);
+	int err = 0;
+	uint32_t prev_val;
+	uint32_t now;
+	uint32_t diff;
+	uint32_t max_rel_val;
+
+	__ASSERT(!check_it_enabled[id](timer),
+		 "Expected that CC interrupt is disabled.");
+
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future).
+	 */
+	now = read(dev);
+	prev_val = get_timer_compare[id](timer);
+	set_timer_compare[id](timer, now);
+	clear_it_flag[id](timer);
+
+	if (absolute) {
+		max_rel_val = top - get_dev_data(dev)->guard_period;
+		irq_on_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+	} else {
+		/* If relative value is smaller than half of the counter range
+		 * it is assumed that there is a risk of setting value too late
+		 * and late detection algorithm must be applied. When late
+		 * setting is detected, interrupt shall be triggered for
+		 * immediate expiration of the timer. Detection is performed
+		 * by limiting relative distance between CC and counter.
+		 *
+		 * Note that half of counter range is an arbitrary value.
+		 */
+		irq_on_late = val < (top / 2);
+		/* limit max to detect short relative being set too late. */
+		max_rel_val = irq_on_late ? top / 2 : top;
+		val = ticks_add(now, val, top);
+	}
+
+	set_timer_compare[id](timer, val);
+
+	/* decrement value to detect also case when val == read(dev). Otherwise,
+	 * condition would need to include comparing diff against 0.
+	 */
+	diff = ticks_sub(val - 1, read(dev), top);
+	if (diff > max_rel_val) {
+		if (absolute) {
+			err = -ETIME;
+		}
+
+		/* Interrupt is triggered always for relative alarm and
+		 * for absolute depending on the flag.
+		 */
+		if (irq_on_late) {
+			set_cc_int_pending(dev, id);
+		} else {
+			get_stm32_config(dev)->ch_data[id].callback = NULL;
+		}
+	} else {
+		enable_it[id](timer);
+	}
+
+	return err;
+}
+
+static int set_alarm(const struct device *dev, uint8_t chan,
+		     const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_stm32_config *config = get_stm32_config(dev);
+	struct counter_stm32_ch_data *chdata = &config->ch_data[chan];
+
+	if (alarm_cfg->ticks >  get_top_value(dev)) {
+		return -EINVAL;
+	}
+
+	if (chdata->callback) {
+		return -EBUSY;
+	}
+
+	chdata->callback = alarm_cfg->callback;
+	chdata->user_data = alarm_cfg->user_data;
+
+	return set_cc(dev, chan, alarm_cfg->ticks, alarm_cfg->flags);
```

Comment:
Nitpick: set_cc can take alarm_cfg directly instead of alarm_cfg->ticks and alarm_cfg->flags

Reply:
Done.